python/pyarrow/parquet.py

```diff
@@ -1172,6 +1190,11 @@ def all_filters_accept(piece):
         self.pieces = [p for p in self.pieces if all_filters_accept(p)]
 
 
+for attribute in _ParquetDatasetMetadata.__slots__:
+    setattr(ParquetDataset, attribute, _metadata_forwarder(attribute))
```

Comment:
Hmm, no, this is making things more bizarre than they already are. I was hoping for a helper object that would expose the current `_make_manifest` and/or `_open_dataset` logic.

Also, if you want to forward a bunch of attributes (at most 5 of them, apparently), I'd suggest writing the properties by hand. You probably don't need to make them writable, either, if the target object is properly initialized.

Reply:
Agreed that a helper object is probably the best approach, but unfortunately I have limited time to spend on this problem. Can we establish a compromise? 

What I can do is:
 * eliminate this funny for-loop and write out the five properties,
 * remove the write-side of the property, which should remove the funny `_metadata_forwarder` function,
 * and change `ParquetDataset.__init__` to write to `self._metadata.the_property` directly

This would also make these metadata properties read-only, which might have some benefits (or drawbacks).

If this is unacceptable, I can remove the `ParquetDataset` changes and leave this PR to just be the changes to `_build_nested_path`, and either you or I can make another ticket to track the `ParquetDataset` reference cycle which will be left not fixed.